packages/firestore/test/integration/api/array_transforms.test.ts

```diff
@@ -0,0 +1,253 @@
+      );
+
+      // wait for initial null snapshot to avoid potential races.
+      const snapshot = await waitForRemoteEvent();
+      expect(snapshot.exists).to.be.false;
+      await test();
+      listenerRegistration();
+    });
+  }
+
+  it('create document with arrayUnion()', async () => {
+    await withTestSetup(async () => {
+      await docRef.set({ array: FieldValue._arrayUnion(1, 2) });
+      expectLocalAndRemoteEvent({ array: [1, 2] });
+    });
+  });
+
+  it('append to array via update()', async () => {
+    await withTestSetup(async () => {
+      await writeInitialData({ array: [1, 3] });
+      await docRef.update({ array: FieldValue._arrayUnion(2, 1, 4) });
+      await expectLocalAndRemoteEvent({ array: [1, 3, 2, 4] });
+    });
+  });
+
+  it('append to array via set(..., {merge: true})', async () => {
+    await withTestSetup(async () => {
+      await writeInitialData({ array: [1, 3] });
+      await docRef.set(
+        { array: FieldValue._arrayUnion(2, 1, 4) },
+        { merge: true }
+      );
+      await expectLocalAndRemoteEvent({ array: [1, 3, 2, 4] });
+    });
+  });
+
+  it('append object to array via update()', async () => {
+    await withTestSetup(async () => {
+      await writeInitialData({ array: [{ a: 'hi' }] });
+      await docRef.update({
+        array: FieldValue._arrayUnion({ a: 'hi' }, { a: 'bye' })
+      });
+      await expectLocalAndRemoteEvent({ array: [{ a: 'hi' }, { a: 'bye' }] });
+    });
+  });
+
+  it('remove from array via update()', async () => {
+    await withTestSetup(async () => {
+      await writeInitialData({ array: [1, 3, 1, 3] });
+      await docRef.update({ array: FieldValue._arrayRemove(1, 4) });
+      await expectLocalAndRemoteEvent({ array: [3, 3] });
+    });
+  });
+
+  it('remove from array via set(..., {merge: true})', async () => {
+    await withTestSetup(async () => {
+      await writeInitialData({ array: [1, 3, 1, 3] });
+      await docRef.set(
+        { array: FieldValue._arrayRemove(1, 4) },
+        { merge: true }
+      );
+      await expectLocalAndRemoteEvent({ array: [3, 3] });
+    });
+  });
+
+  it('remove object from array via update()', async () => {
+    await withTestSetup(async () => {
+      await writeInitialData({ array: [{ a: 'hi' }, { a: 'bye' }] });
+      await docRef.update({ array: FieldValue._arrayRemove({ a: 'hi' }) });
+      await expectLocalAndRemoteEvent({ array: [{ a: 'bye' }] });
+    });
+  });
+
+  /**
+   * Unlike the withTestSetup() tests above, these tests intentionally avoid
+   * having any ongoing listeners so that we can test what gets stored in the
+   * offline cache based purely on the write acknowledgement (without receiving
+   * an updated document via watch). As such they also rely on persistence
+   * being enabled so documents remain in the cache after the write.
+   */
+  (persistence ? describe : describe.skip)('Server Application: ', () => {
+    it('set() with no cached base doc', async () => {
```

Comment:
Maybe I am missing something here - but could we not simplify these tests by disabling the network and querying from snapshots purely from cache?

Reply:
Yeah, that makes sense. I was thinking you could take your knowledge of the codebase into account and test `applyToRemoteDocument` by calling `applyToLocalView` (shrug).

Reply:
Mostly I was nervous about the fact that the server sends us a null transform result and I wanted to make sure that we correctly ignore that and perform the local transform instead.  So I wanted to test that end-to-end.  It's possibly overkill or there are better ways to test it, but I'm inclined to just stick with these tests.